reconcile/terraform_resources_cloudflare.py

```diff
@@ -0,0 +1,283 @@
+        name = i.integration
+        if name.replace("-", "_") == QONTRACT_INTEGRATION:
+            # we have to ensure the bucket key(file) is unique across
+            # Cloudflare accounts to support running per-account
+            bucket_key = f"{QONTRACT_INTEGRATION}-{cf_acct.name}.tfstate"
+            bucket_name = tf_state.bucket
+            bucket_region = tf_state.region
+            break
+
+    if bucket_name and bucket_key and bucket_region:
+        backend_config = TerraformS3BackendConfig(
+            aws_acct_creds["aws_access_key_id"],
+            aws_acct_creds["aws_secret_access_key"],
+            bucket_name,
+            bucket_key,
+            bucket_region,
+        )
+    else:
+        raise ValueError(f"No state bucket config found for account {aws_acct.name}")
+
+    return backend_config
+
+
+def get_resources(
+    query_data: TerraformResourcesCloudflareQueryData,
+) -> list[NamespaceTerraformProviderResourceCloudflareV1]:
+    """Get all Cloudflare V1 resources from the Cloudflare query data"""
+    return [
+        res
+        for namespace in filter_null(query_data.namespaces)
+        for res in filter_null(namespace.external_resources)
+        if isinstance(res, NamespaceTerraformProviderResourceCloudflareV1)
+    ]
+
+
+def build_clients(
+    settings: dict[str, Any],
+    query_data: TerraformResourcesCloudflareQueryData,
+) -> list[tuple[str, TerrascriptCloudflareClient]]:
+    clients = []
+    for extres in get_resources(query_data):
+        cf_acct = extres.provisioner
+        cf_acct_config = create_cloudflare_account_config(settings, cf_acct)
+
+        aws_acct = cf_acct.terraform_state_account
+        aws_backend_config = create_backend_config(settings, aws_acct, cf_acct)
+
+        ts_config = create_cloudflare_terrascript(
+            cf_acct_config,
+            aws_backend_config,
+            cf_acct.provider_version,
+        )
+
+        ts_client = TerrascriptCloudflareClient(ts_config)
+        clients.append((cf_acct.name, ts_client))
+    return clients
+
+
+def get_github_file(repo: str, path: str, ref: str) -> str:
+    settings = queries.get_app_interface_settings()
+    gh_instance = queries.get_github_instance()
+    gh = GithubApi(gh_instance, repo, settings)
+    content = gh.get_file(path, ref)
+    if content is None:
+        raise ValueError(
+            f"Could not retrieve Github file content at {repo} "
+            f"for file path {path} at ref {ref}"
+        )
+    return content.decode("utf-8")
+
+
+def build_specs(
+    query_data: TerraformResourcesCloudflareQueryData,
+) -> list[ExternalResourceSpec]:
+
+    specs = []
+    for extres in get_resources(query_data):
+        provisioner_name = extres.provisioner.name
+        for res in extres.resources or []:
+            res = cast(NamespaceTerraformResourceCloudflareZoneV1, res)
+            zone_identifier = safe_resource_id(res.zone)
+            specs.append(
+                ExternalResourceSpec(
```

Comment:
I initially had everything under a big CloudflareZoneTerrascriptResource and a single ExternalResourceSpec here, but that made the code complicated with everything under the same class. See here https://github.com/app-sre/qontract-reconcile/pull/2691/commits/51db5f554f25f9e62e2e920dd09af8abc7c3f478#diff-37ffa73fa5362db9454a2f178916e524da8157290c659d1e2b0aae673dc64977R64

Or perhaps I am misunderstanding your concern?

Reply:
I was thinking that the desire to structure code in a way that's easier to read and maintain shouldn't require changes to how we translate objects from the GraphQL representation to the Python objects meant to represent them? Another way we could achieve this would be:

```python
class CloudflareZoneTerrascriptResource(TerrascriptResource):

    def _create_zone_records(zone_records) -> list[cloudflare_record]:
         (code related to creating zone records)
         
    def _create_zone_workers(zone_workers) -> list[cloudflare_worker_script]:
        (code related to creating zone workers)

    def populate(self) -> list[Union[Resource, Output]]:
        (call the private methods here)
```

We'd have private methods to logically break up the code and make it easier to follow. This still results in one class that would be quite small and only has one public method, but doesn't require us to translate a single `ExternalResourceSpec` into others that don't actually exist as resources in our queries. I might be missing something here, but it just seems odd that 1 resource in GraphQL would become > 1 `ExternalResourceSpec` when they're currently 1:1 everywhere else (I don't see `ExternalResourceSpec` being created anywhere else besides `get_external_resource_specs`).

Happy to discuss further if I can make this concern clearer. I'm also happy to get another opinion on this one if you think this isn't a major concern.

`[side note]`: This is also all feels like a symptom of having an existing schema (from the Terraform provider) and trying to lay another schema (qontract-schemas) on top of it. I'm guessing zone was defined as a single resource because other methods of defining relationships aren't as straight-forward as it would be in Terraform. I only mention this because it feels like another +1 for additional discussions about Terraform modules and skipping the additional abstraction on top of Terraform (there could be some downsides as well).

Reply:
the whole method is gone and code was refactored to use `get_external_resource_specs`